### PR TITLE
Fix URL trailing symbols

### LINKED ISSUE #10898: URL incorrectly includes trailing punctuation
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Signal for Android App takes any punctuation following an inline URL as part of the URL. This makes the URL not work and it must be hand edited (to strip off any trailing punctuation) in the address bar of the browser after clicking/touching.

Extra info: Signal for Windows does the correct thing.

### Steps to reproduce
One specific example:
- share location with Google Maps via a signal message
- Signal for Android of the receiver includes the punctuation following the URL as part of the clickable URL itself

**Actual result:** URL includes punctuation directly following the URL
**Expected result:** URL does not contain punctuation directly following the URL

### Screenshots
This is a screenshot using Signal for Android App. Notice the final period is part of the URL (underlined):

image

This is a screenshot using Signal for Windows of the very same message. Here, the final period is not part of the URL (not underlined), which is correct:

image

### Device info
**Device:** Nokia 9 Pureview
**Android version:** 10
**Signal version:** 5.3.7.1

### LINKED ISSUE #11310: A trailing dash/hyphen of an URL is not "hyperlinked"
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
When I send an URL with a trailing hyphen, this hyphen is missing in the hyperlink added by Signal on this URL.
Example of a sent message : 
> https://fr.ulule.com/sapins-barbus-la-bd-/

The hyperlink goes to https://fr.ulule.com/sapins-barbus-la-bd

### Steps to reproduce
- send a message, containing an URL with a trailing dash
- the trailing dash is missing in the hyperlink

**Actual result:** Because the hyperlink is wrong, it could generate a 404 error when the user click on the URL
**Expected result:** The hyperlink should be faithful to the URL, and contain all the characters indicated

### Screenshots
Screenshot_20210517-125638_Signal

### Device info

**Device:** Samsung Galaxy S7
**Android version:** 8.0.0
**Signal version:** 5.11.3

### PR DESCRIPTION
**Pull request by ylpoonlg**
### First time contributor checklist

- [x] I have read [how to contribute](https://github.com/signalapp/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [x] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [x] I am following the [Code Style Guidelines](https://github.com/signalapp/Signal-Android/wiki/Code-Style-Guidelines)
- [x] I have tested my contribution on these devices:
 * Samsung A9000, Android 6.0.1
 * Samsung A9200, Android 10
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description

Fixes #10898, fixes #11310.
Includes trailing punctuation to the URLSpan of messageBody.
It looks for the first illegal url character after the original span and extend the span so that anything in between is included.

(Reopened #12308 with different branch.)

### CODE REVIEW

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java [hunk 1]
```diff
@@ -1402,12 +1406,26 @@ private void linkifyMessageBody(@NonNull Spannable messageBody,
...
+      URLSpan[] urlSpans = messageBody.getSpans(0, messageBody.length(), URLSpan.class);
+
+      for (URLSpan urlSpan : urlSpans) {
+        int     start = messageBody.getSpanStart(urlSpan);
+        int     end   = messageBody.getSpanEnd(urlSpan);
+
+        Matcher matcher = NOT_URL_PATTERN.matcher(messageBody.toString().substring(end));
+        if (matcher.find()) {
+          int newEnd = end + matcher.start();
+          URLSpan newSpan = new URLSpan(messageBody.toString().substring(start, newEnd));
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java:1418]: formatting:

```
int     newEnd  =
URLSpan newSpan =
```

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java [hunk 2]
```diff
@@ -1402,12 +1406,26 @@ private void linkifyMessageBody(@NonNull Spannable messageBody,
     boolean hasLinks    = LinkifyCompat.addLinks(messageBody, shouldLinkifyAllLinks ? linkPattern : 0);
 
     if (hasLinks) {
-      Stream.of(messageBody.getSpans(0, messageBody.length(), URLSpan.class))
+      URLSpan[] urlSpans = messageBody.getSpans(0, messageBody.length(), URLSpan.class);
+
+      for (URLSpan urlSpan : urlSpans) {
+        int     start = messageBody.getSpanStart(urlSpan);
+        int     end   = messageBody.getSpanEnd(urlSpan);
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java:1413]: Formatting

```
int start = 
int end   =
```

File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java [hunk 3]
```diff
@@ -1402,12 +1406,26 @@ private void linkifyMessageBody(@NonNull Spannable messageBody,
     boolean hasLinks    = LinkifyCompat.addLinks(messageBody, shouldLinkifyAllLinks ? linkPattern : 0);
 
     if (hasLinks) {
-      Stream.of(messageBody.getSpans(0, messageBody.length(), URLSpan.class))
+      URLSpan[] urlSpans = messageBody.getSpans(0, messageBody.length(), URLSpan.class);
```
Comment (alex-signal) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationItem.java:1409]: Definitely *not* something you need to do in this PR but this linkifying logic feels like a great candidate for a set of unit tests.

**Review by alex-signal — CHANGES_REQUESTED** [review]

**Comment by alex-signal** [comment]
This is looking good! For my own education, and use in Unit tests (that I'll hopefully write once this is pulled in) could you add a few examples of URLs that were previously broken that this change fixes? (Just in a github comment, no need to do so in code)

**Comment by ylpoonlg** [comment]
[https://fr.ulule.com/sapins-barbus-la-bd-/](https://fr.ulule.com/sapins-barbus-la-bd-/)
[https://www.instagram.com/tv/CfImYdngccQ/?igshid=YmMyMTA2M2Y=](https://www.instagram.com/tv/CfImYdngccQ/?igshid=YmMyMTA2M2Y=)

Apparently, the problem only exist when there is a symbol (e.g. .,/+-=?) at the end followed by a space or a newline character.

**Comment by alex-signal** [comment]
thanks! I've added these as test cases.

**Review by alex-signal — APPROVED** [review]

**Pull request closed without merging**